core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSigner.java

```
@@ -0,0 +1,110 @@
+        boolean doubleUrlEncode = signRequest.requireProperty(AwsV4HttpSigner.DOUBLE_URL_ENCODE, true);
+        boolean normalizePath = signRequest.requireProperty(AwsV4HttpSigner.NORMALIZE_PATH, true);
+        AuthLocation authLocation = signRequest.requireProperty(AUTH_LOCATION, AuthLocation.HEADER);
+        Duration expirationDuration = signRequest.requireProperty(EXPIRATION_DURATION, null);
+        boolean isPayloadSigning = signRequest.requireProperty(PAYLOAD_SIGNING_ENABLED, true);
+
+        Instant signingInstant = signingClock.instant();
+        AwsCredentialsIdentity credentials = sanitizeCredentials(signRequest.identity());
+        CredentialScope credentialScope = new CredentialScope(regionName, serviceSigningName, signingInstant);
+        Checksummer checksummer = Checksummer.create();
+        Function<V4Properties, V4RequestSigner> requestSigner = properties -> V4RequestSigner.create(null);
+        V4PayloadSigner payloadSigner = V4PayloadSigner.create();
+
+        if (authLocation == AuthLocation.HEADER) {
+            requestSigner = V4RequestSigner::header;
+        } else if (authLocation == AuthLocation.QUERY_STRING) {
+            requestSigner = V4RequestSigner::query;
+            if (expirationDuration != null) {
+                if (expirationDuration.compareTo(PRESIGN_URL_MAX_EXPIRATION_DURATION) > 0) {
+                    throw new IllegalArgumentException(
+                        "Requests that are pre-signed by SigV4 algorithm are valid for at most 7" +
+                        " days. The expiration duration set on the current request [" + expirationDuration + "]" +
+                        " has exceeded this limit."
+                    );
+                }
+                requestSigner = properties -> V4RequestSigner.presigned(properties, expirationDuration);
+            }
+        }
+
+        if (!isPayloadSigning) {
+            checksummer = new PrecomputedChecksummer(() -> "UNSIGNED-PAYLOAD");
```

Does this need different handling when protocol is `http`?

---

I hope there's an integ test that with `http` to make sure this works end-to-end when we have the new signers fully integrated. If not, we should add one.

---

Yeah, good call-out. We should add a check here to use payload-signing if it's http, or we can throw unsupported in that case too: "Payload signing must be enabled if http is used."